src/learner.cc

```diff
@@ -332,8 +339,15 @@ class LearnerImpl : public Learner {
       }
     }
     {
-      // Write `predictor`, `gpu_id` parameters as extra attributes
-      for (const auto& key : std::vector<std::string>{"predictor", "gpu_id"}) {
+      std::vector<std::string> saved_params{"predictor", "n_gpus", "gpu_id"};
+      // check if rabit_bootstrap_cache were set to non zero before adding to checkpoint
+      if (cfg_.find("rabit_bootstrap_cache") != cfg_.end() &&
```

Comment:
Have you tested backward compatibility?

Reply:
if we don't explicit set rabit_bootstrap_cache to 1, yes, it was tested read/write old format with new code path.

If we set rabit_bootstrap_cache to 1, and read old format or rabit_bootstrap_cache=0, yes, it passed all tests except failed xgb worker recovery. That is because additional parameters were empty exactly same format as we have before this pr.

In both cases, it can load old checkpoint format and pass tests.